mem/ctrl/ContextSwitchRTL.py

```diff
@@ -0,0 +1,95 @@
+"""
+==========================================================================
+ContextSwitchRTL.py
+==========================================================================
+Records/resumes progress (itertion) for functional unit (FU).
+
+Author : Yufei Yang
+  Date : Aug 11, 2025
+"""
+from pymtl3.stdlib.primitive import *
+from ...lib.basic.val_rdy.ifcs import ValRdyRecvIfcRTL as RecvIfcRTL
+from ...lib.basic.val_rdy.ifcs import ValRdySendIfcRTL as SendIfcRTL
+from ...lib.cmd_type import *
+from ...lib.opt_type import *
+from ...lib.util.common import *
+
+class ContextSwitchRTL(Component):
+
+  def construct(s, CgraPayloadType, DataType, CtrlType, num_status = 3):
+
+    # Constant
+    StatusType = mk_bits(clog2(num_status))
+
+    # Interface
+    s.recv_cmd = RecvIfcRTL(CgraPayloadType)
+    s.recv_opt = RecvIfcRTL(CtrlType)
+    s.progress_in = RecvIfcRTL(DataType)
+    s.progress_out = SendIfcRTL(DataType)
+   
+    # Component
+    s.progress_reg = Reg(DataType)
+    s.status_reg = Reg(StatusType)
+    #s.condition = Wire(Bits3)
+    # How to create a wire that can be bit sliced?
+    s.condition_2 = Wire(b1)
+    s.condition_1 = Wire(b1)
+    s.condition_0 = Wire(b1)
+
+    @update
+    def update_msg():
+      # Update condition.
+      # condition = 3'b111 (3'd7): Records the progress.
+      # condition = 3'b001 (3'd1): Resumes the progress.
+      s.condition_2 @= (s.progress_reg.out == DataType(0))
+      s.condition_1 @= (s.status_reg.out == StatusType(3))
+      s.condition_0 @= s.recv_opt.val & (s.recv_opt.msg.operation == OPT_PHI_CONST)
+
+      # Updates recv_opt.
+      s.recv_opt.rdy @= 1
+
+      # Updates recv_cmd
+      s.recv_cmd.rdy @= 1
+
+      # Updates the status register.
+      if (s.recv_cmd.val & (s.recv_cmd.msg.cmd == CMD_PAUSE)):
+        # PAUSING: 2'b11
+        s.status_reg.in_ @= StatusType(3)
```

Comment:
this `s.status_reg` is updated like a combinational logic, rather than "on_edge", I can't remember it clearly, but you mean as long as we set its `.in_`, the content would be updated/available in next cycle?

Why not just declare them as `Wire`, and update them in an `@update_ff`? They would be totally under your control, and will be synthesized as registers.